server/encryptionkm/key_manager.go

```diff
@@ -14,41 +14,469 @@
+	}
+	// Load encryption keys from storage.
+	_, err = m.loadKeys()
+	if err != nil {
+		return nil, err
+	}
+	// Start periodic check for keys change and rotation key if needed.
+	go m.startBackgroundLoop(ctx, m.mu.keysRevision)
+	return m, nil
+}
+
+func (m *KeyManager) startBackgroundLoop(ctx context.Context, revision int64) {
+	// Create new context for the loop.
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	// Setup key dictionary watcher
+	watcher := clientv3.NewWatcher(m.etcdClient)
+	watchChan := watcher.Watch(ctx, EncryptionKeysPath, clientv3.WithRev(revision))
+	watcherEnabled := true
+	defer watcher.Close()
+	// Check data key rotation every min(dataKeyRotationPeriod, keyRotationCheckPeriod).
+	checkPeriod := m.dataKeyRotationPeriod
+	if keyRotationCheckPeriod < checkPeriod {
+		checkPeriod = keyRotationCheckPeriod
+	}
+	ticker := time.NewTicker(checkPeriod)
+	defer ticker.Stop()
+	// Loop
+	for {
+		select {
+		// Reload encryption keys updated by PD leader (could be ourselves).
+		case resp := <-watchChan:
+			if resp.Canceled {
+				// If the watcher failed, we fallback to reload every 10 minutes.
+				log.Warn("encryption key watcher canceled")
+				watcherEnabled = false
+				continue
+			}
+			for _, event := range resp.Events {
+				if event.Type != mvccpb.PUT {
+					log.Warn("encryption keys is deleted unexpectedly")
+					continue
+				}
+				m.mu.Lock()
+				_, err := m.loadKeysFromKV(event.Kv)
+				m.mu.Unlock()
+				if err != nil {
+					log.Warn("fail to get encryption keys from watcher result", zap.Error(err))
+				}
+			}
+			m.helper.eventAfterReloadByWatcher()
+			// Check data key rotation in case we are the PD leader.
+		case <-m.helper.tick(ticker):
+			m.mu.Lock()
+			err := m.rotateKeyIfNeeded(false /*forceUpdate*/)
+			if err != nil {
+				log.Warn("fail to rotate data encryption key", zap.Error(err))
+			}
+			// Fallback mechanism to reload keys if watcher failed.
+			if !watcherEnabled {
+				_, err = m.loadKeys()
+				if err != nil {
+					log.Warn("fail to reload keys after watcher failed", zap.Error(err))
+				}
+			}
+			m.mu.Unlock()
+			m.helper.eventAfterTicker()
+			// Server shutdown.
+		case <-ctx.Done():
+			log.Info("encryption key manager is closed")
+			return
+		}
+	}
+}
+
+// loadKeysFromKV reload keys from etcd result.
+// Require mu lock to be held.
+func (m *KeyManager) loadKeysFromKV(
+	kv *mvccpb.KeyValue,
+) (*encryptionpb.KeyDictionary, error) {
+	// Sanity check if keys revision is in order.
+	// etcd docs indicates watcher event can be out of order:
+	// https://etcd.io/docs/v3.4.0/learning/api_guarantees/#isolation-level-and-consistency-of-replicas
+	if kv.ModRevision <= m.mu.keysRevision {
+		return m.getKeys(), nil
+	}
+	keys, err := loadKeysFromKV(kv)
+	if err != nil {
+		return nil, err
+	}
+	m.mu.keysRevision = kv.ModRevision
+	m.keys.Store(keys)
+	log.Info("reloaded encryption keys", zap.Int64("revision", kv.ModRevision))
+	return keys, nil
+}
```

Comment:
It seems the `keysRevision` is visited and updated withou holding mutex here.

Reply:
loadKeysFromKV are supposed to be called with mutex held.